lxd/instance/drivers/driver_qemu.go

```diff
@@ -375,7 +376,11 @@ func (d *qemu) getAgentClient() (*http.Client, error) {
 		return nil, err
 	}
 
-	vsockID := d.vsockID() // Default to using the vsock ID that will be used on next start.
+	// Default to using the vsock ID that will be used on next start.
+	vsockID, err := d.vsockID()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Can we swap the ordering of this part with the volatile.vsock_id detection below such that if the volatile key is present and valid we don't need to call d.vsockID() at all?

Reply:
Oh actually you've added the volatile.vsock_id logic to d.vsockID() in the right order, so we dont need the duplicated logic below right?

Reply:
On further consideration I think we should just use the value from volatile.vsock_id here and not try to generate a new one.

This worked OK when it was predictable without trying to listen, but now its not, there's no point in calling d.vsockID() here.

Reply:
I then think that the d.vsockID() function should try and check that the current volatile.vsock_id is free, and use that if so, otherwise try and find a new free one. This way if the volatile.vsock_id becomes out of date when the instance is stopped, it will recover gracefully.

Reply:
Since `d.vsockID()` can return the `volatile.vsock_id`, I have replaced it completely so that only this function returns this config setting. 
Since there is now also a check for the existing Context ID being already acquired, `d.vsockID()` would anyway just return the config setting without picking a new ID.

The comments are moved accordingly.